compiler/resolution/fixupExports.cpp

```diff
@@ -0,0 +1,425 @@
+      if (wrapper == NULL) { wrapper = createWrapper(fn); }
+      VarSymbol* tmp = fixupFormal(wrapper, i);
+      INT_ASSERT(tmp != NULL);
+      tmps.push_back(tmp);
+    } else {
+      // Push back a NULL sentry value for unconverted formals.
+      tmps.push_back(NULL);
+    }
+  }
+
+  if (needsFixup(fn->retType)) {
+    validateReturnIntent(fn);
+    if (wrapper == NULL) { wrapper = createWrapper(fn); }
+    changeRetType(wrapper);
+  }
+
+  // If a wrapper hasn't been made yet, there's nothing to do.
+  if (wrapper == NULL) { return; }
+
+  insertUnwrappedCall(wrapper, fn, tmps);
+
+  wrapperMap[wrapper] = fn;
+
+  return;
+}
+
+static bool needsFixup(ArgSymbol* as) {
+  return needsFixup(as->type);
+}
+
+static bool needsFixup(Type* t) {
+  if (t == dtString) {
+    return true;
+  }
+
+  return false;
+}
+
+static void validateFormalIntent(FnSymbol* fn, int idx) {
+  ArgSymbol* as = fn->getFormal(idx);
+  Type* t = as->type;
+
+  //
+  // TODO: If we ever add more types to these fixup routines, we really ought
+  // to put these conditions in tables.
+  //
+  if (t == dtString) {
+    IntentTag tag = as->intent;
+
+    // TODO: After resolution, have abstract intents been normalized?
+    if (tag != INTENT_CONST &&
+        tag != INTENT_CONST_REF &&
+        tag != INTENT_BLANK) {
+      SET_LINENO(fn);
+      USR_FATAL(as, "Formal %s of type \'%s\' in exported routine \'%s\' may "
+                    "only have the %s",
+                as->name, t->name(), fn->userString,
+                intentDescrString(INTENT_CONST_REF));
+    }
+  }
+
+  return;
+}
+
+static void validateReturnIntent(FnSymbol* fn) {
+  Type* t = fn->retType;
+
+  if (t == dtString) {
+    RetTag tag = fn->retTag;
+
+    if (tag != RET_VALUE) {
+      SET_LINENO(fn);
+      USR_FATAL(fn, "Return type %s of exported routine %s may only have "
+                    "return intent %s",
+                t->name(), fn->userString,
+                retTagDescrString(RET_VALUE));
+    }
+  }
+
+  return;
+}
+
+static FnSymbol* createWrapper(FnSymbol* fn) {
+  FnSymbol* result = fn->copy();
+  fn->defPoint->insertAfter(new DefExpr(result));
+  // TODO: Is there a better way to empty the block body?
+  result->body->replace(new BlockStmt());
+  INT_ASSERT(fn->hasFlag(FLAG_EXPORT));
+  fn->removeFlag(FLAG_EXPORT);
+  return result;
+}
+
+static Type* getTypeForFixup(Type* t, bool ret) {
+  if (t == dtString) {
+    Type* result = ret ? getCharPtrType() : dtStringC;
+    return result;
+  } else {
+    INT_FATAL("Unsupported type for formal in: %s", __FUNCTION__);
+  }
+
+  // Should never reach here.
+  return NULL;
+}
+
+//
+// TODO: Can insert non-copying versions when we are doing multi-locale
+// interop.
+//
+static const char* getConversionCallName(Type* srctype, Type* dsttype) {
+  const char* result = NULL;
+
+  // TODO: We should probably maintain this in a table or the like.
+  if (srctype == dtString && dsttype == dtStringC) {
```

Comment:
An alternative to this is to have a single function and rely on formal constraints to resolve correctly. For example:
```chpl
proc chpl__exportWrapper(val:string, type ret:c_string) : ret { ... }
proc chpl__exportWrapper(val:string, type ret:c_ptr(c_char)) : ret { ... }
proc chpl__exportWrapper(val:c_string, type ret:string) : ret { ... }
proc chpl__exportWrapper(val:c_ptr(c_char), type ret:string) : ret { ... }
```

That way you can just have the one name.

Reply:
I was wondering how I might go about doing this. That's awesome, thanks!

Reply:
For this idea, what AST would I emit for the `type ret:c_string...` part of the call?

Reply:
I think you'd want to create a new temporary, add ``FLAG_TYPE_VARIABLE``, and set its type to ``c_string``.